surface/include/pcl/surface/3rdparty/poisson4/octree_poisson.hpp

```diff
@@ -794,7 +794,8 @@ namespace pcl
     long long _InterleaveBits( int p[3] )
     {
       long long key = 0;
-      for( int i=0 ; i<32 ; i++ ) key |= ( ( p[0] & (1<<i) )<<(2*i) ) | ( ( p[1] & (1<<i) )<<(2*i+1) ) | ( ( p[2] & (1<<i) )<<(2*i+2) );
+      long long _p[3] = {p[0],p[1],p[2]};
```

Comment:
Would it make sense to deprecate current `_InterleaveBits` in favor of `std::int32_t` and `std::int64_t` versions? Maybe send the non-deprecated one in a detail namespace

Reply:
I think so, since long long, and int are system dependent, right? Moving to std::intXX_t would be safer. But would we need a 32-bit version for this application?

Reply:
>  long long, and int are system dependent, right?

Only `int` is system dependent. For 64 bit systems, `long long` is compulsorily 64 bits (LLP64, LP64, ILP64, SILP64)

### Clarification
I meant `std::int64_t` as return value (aka `long long int`) and `std::int32_t` AND `std::int64_t` as input values. I noticed `one` singular form instead of `ones`. Sorry for the confusion. This is needed since some platforms have `int` as 64 bits (ILP64, SILP64)

> But would we need a 32-bit version for this application?

**Long story short**: We do compile using a 32 bit compiler, but don't worry about this. 

**Long story**: 32 bit compiler wouldn't affect anything since 64 bit arithmetic is simulated in software by the compiler using 2 32-bit words. 64 bits are guaranteed by the language spec contingent on a implementation in `stdint.h` (along with compiler support for in software simulation) [sample header](https://www.boost.org/doc/libs/1_37_0/boost/cstdint.hpp). If 64 bits are missing, `long long` by itself becomes a compiler error as per C99 standard. So, we're ok changing to fixed integer sizes.

Reply:
Ok, I got it now. I agree with having the std::int64_t as both input and output, although it would change the more than just adding it and deprecating the old one. 

Regarding the std::int32_t for input, which would be similar to what we have now, should we continue constructing a std::int64_t inside the function? That is: 

`std::int64_t _p[3] = {p[0],p[1],p[2]};`

Reply:
Yeah. And then call the 64-bit version. Seems like no adverse effect on [godbolt](https://godbolt.org/z/n2p9qq)

Reply:
Ok, cool. The issue will be naming the new functions since `std::int32_t == int` so one function would be a redefinition of the deprecated.

> Maybe send the non-deprecated one in a detail namespace

I guess this is a solution, where should this namespace sit? Should it be in-place?

```cpp
namespace pcl
{
  namespace poisson
  {
    .
    .
    .
    namespace detail
    {
      std::int64_t _InterleaveBits( std::int64_t p[3] ) ...
      std::int64_t _InterleaveBits( std::int32_t p[3] ) ...
    }
    [[deprecated("use detail::_InterleaveBits() overload that takes std::int32_t or std::int64_t instead.")]]
    long long _InterleaveBits( int p[3] ) ...
    .
    .
  }
}
```

Reply:
Yeah. the deprecated one just calls the non deprecated one (magically works). And so do the actual call-sites. I think there are only 2 or 3 of them.

Maybe rename the functions in detail simply `interleaveBits`? The sole reason for that naming could be making it weird for the public API.

I wish we could say `private API use. Will be removed in the future`. We should not tell people to use a non-public API function. @taketwo what do you think?

Reply:
> Maybe rename the functions in detail simply `interleaveBits`?

:+1:

> I wish we could say something like private API use. Will be removed in the future. We should not tell people to use a non-public API function. @taketwo what do you think?

Not sure I understand the question. For me, everything in `detail` namespace is mmm... _implementation detail_, and as such is subject to change without notice.

Reply:
This function isn't in detail namespace, but in the `hpp` aka implementation file. It's implicit that it was not for public consumption. Also, the naming is hinting towards that.

Reply:
Sorry, I was away this weekend. Should we keep it as it is then? Since this was merged already, I guess we assume this function as an implementation detail so we don't need to deprecate it or create alternatives, right?